cdap-common/src/main/java/co/cask/cdap/internal/io/DatumWriterGenerator.java

```diff
@@ -404,9 +404,15 @@ private void encodeSimple(GeneratorAdapter mg, TypeToken<?> type, Schema schema,
     TypeToken<?> encodeType = type;
     mg.loadArg(encoder);
     mg.loadArg(value);
-    if (Primitives.isWrapperType(encodeType.getRawType())) {
+
+    boolean isPrimitiveBoxType = Primitives.isWrapperType(encodeType.getRawType());
+    if (encodeType.getRawType().isPrimitive() || isPrimitiveBoxType) {
+      // Unwrap it if it is boxed type, otherwise it won't change the encodeType.
```

Comment:
The comment is intended for the unwrapping of boxed type to primitive type. Line 413 is for generating an unbox statement in the bytecode.